### PR TITLE
Sprint 9 Day 3 Part 1: i++1/i--1 Arithmetic Indexing Grammar & Semantic Handlers

### PR DESCRIPTION
**Pull request by jeffreyhorn**
## Overview

Implements Sprint 9 Day 3 Part 1: Grammar changes, semantic handlers, and comprehensive unit tests for lead/lag arithmetic indexing (i++1, i--2, i+j, etc.).

## Grammar Changes

- Added tokens: `CIRCULAR_LEAD (++)`, `CIRCULAR_LAG (--)`
- New rules:
  - `index_list`: Supports both plain identifiers and lag/lead expressions
  - `index_expr`: Base identifier with optional lag/lead suffix
  - `lag_lead_suffix`: Circular lead/lag (++/--) or linear (+/-)
  - `offset_expr`: Numeric or variable offset
- Context separation: `index_list` for references, `id_list` for declarations
- No conflicts: Token-level disambiguation (++ matches before +)

## IR Representation

- New node: `IndexOffset(base: str, offset: Expr, circular: bool)`
- Examples:
  - `i++1` → `IndexOffset('i', Const(1), circular=True)`
  - `i--2` → `IndexOffset('i', Const(-2), circular=True)`
  - `i+j` → `IndexOffset('i', SymbolRef('j'), circular=False)`
- Exported in `src/ir/__init__.py`
- Updated `VarRef`, `ParamRef`, `MultiplierRef` to support `tuple[str | IndexOffset, ...]`

## Semantic Handlers

- `_process_index_expr()`: Process single index with optional lag/lead
- `_process_index_list()`: Process comma-separated indices
- Updated `_extract_indices()`: Handle index_expr nodes
- Updated `symbol_indexed`: Use _process_index_list for refs
- Updated `bound_indexed`: Support lag/lead in bounds
- Lag conversion: i--2 → offset=Const(-2)

## Comprehensive Unit Tests ✅

**29 tests** in `tests/unit/test_arithmetic_indexing.py` covering:

### Basic Patterns (7 tests)
- Circular lead: `i++1`
- Circular lag: `i--2`
- Linear lead: `i+1`
- Linear lag: `i-3`
- Variable offsets: `i+j`, `i--j`, `i-j`

### Edge Cases (3 tests)
- Zero offset (`i++0`)
- Large offsets (`i++10`, `i--100`)
- Small sets (2 elements)

### Multi-Dimensional (4 tests)
- 2D with both lead: `x(i++1, j++1)`
- 2D mixed lead/lag: `x(i++1, j--1)`
- 3D all lead: `x(i++1, j++2, k++3)`
- Mixed plain and offset: `x(i, j++1)`

### Equation Indexing (3 tests)
- Circular lead in equations
- Multiple offsets in same equation
- Variable offset in equations

### Bound Indexing (3 tests)
- Lower bound with lead: `x.lo(i++1)`
- Upper bound with lag: `x.up(i--1)`
- Fixed bound with variable offset

### Complex Expressions (3 tests)
- Multiple references: `x(i++1) + x(i--1)`
- Arithmetic with offsets
- Function calls with offsets

### Real-World Patterns (2 tests)
- himmel16.gms pattern
- Inventory balance pattern

### Error Handling (1 test)
- Parameter assignment with IndexOffset (should fail)

### Backward Compatibility (3 tests)
- Plain single index still works
- Plain multi-index still works
- Mixed plain and offset indices

**All 29 tests pass ✅**

## Test Updates

- Updated 2 tests: Expected failures now succeed (i++1 supported)
- All 1390 tests pass ✅

## Regression Validation

- ✅ Existing 4 models parse: mhw4d, rbrock, mathopt1, trig
- ✅ Grammar tests: i++1, i--2, i+j, i-3, multi-dimensional
- ✅ Quality checks: typecheck, lint, format, test

## Implementation Highlights

- Token-level disambiguation (longest match wins)
- Context-aware parsing (indexing vs arithmetic)
- Backward compatible (plain identifiers still work)
- No grammar conflicts
- Comprehensive test coverage for all operators and edge cases

## Next Steps

- Day 4: Validate himmel16.gms parses (95%+ unlock probability)
- Part 2: Implement solver code generation for IndexOffset

## References

- Sprint 9 PLAN.md Day 3 (lines 340-463)
- PREP Task 3 (i++1 research)
- KNOWN_UNKNOWNS.md (9.1.1-9.1.3)

Ready for review!

### CODE REVIEW

File: tests/unit/test_arithmetic_indexing.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+
+    def test_parameter_assignment_with_lead_lag_rejected(self):
+        """Test that lead/lag in parameter assignments is rejected."""
+        source = """
+Set i /1*5/;
+Parameter x(i);
+x(i++1) = 10;
+"""
+        # Parameter assignments don't support lead/lag offsets
+        with pytest.raises(Exception):  # Should raise ParserSemanticError
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply tests/unit/test_arithmetic_indexing.py]: Fixed in bee8897. Changed to `pytest.raises(ParserSemanticError)` and imported ParserSemanticError from src.ir.parser. Now validates the correct exception type.

File: tests/unit/test_arithmetic_indexing.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
+"""Unit tests for arithmetic indexing (i++1, i--1, i+j) - Sprint 9 Day 3.
+
+Tests focus on equation definitions and bounds, which are the primary use cases
+for lead/lag indexing in GAMS (parameter assignments don't support IndexOffset).
+"""
+
+import pytest
+
+from src.ir.parser import parse_model_text
+from src.utils.errors import ParseError
```
[the remark this answers is not in the document]
Comment (jeffreyhorn) [reply tests/unit/test_arithmetic_indexing.py]: Fixed in bee8897. Removed unused ParseError import - it was replaced with ParserSemanticError import when fixing the pytest.raises issue.
